### PR TITLE
Pf 3474 date picker scroll bug

### PR DESCRIPTION
**Pull request by JohnStuartBarrie**
# Description
Datepicker Placement on Scroll Bug

# Screenshots / Gifs
date scroll problems

### CODE REVIEW

File: src/components/date/date.js [hunk 1]
```diff
@@ -167,6 +168,17 @@ class Date extends React.Component {
...
   }
 
+  getDOMNode() {
+    return ReactDOM.findDOMNode(this);
+  }
+
+  getChildContext() {
+    return { 
+      parentDOMNode: this.getDOMNode.bind(this),
+      onResposition: this.updateDatePickerPosition.bind(this)
```
Comment (andrewjtait) [inline src/components/date/date.js]: you don't need to use context for these, just pass them as props

File: src/components/date/date.js [hunk 2]
```diff
@@ -577,7 +576,7 @@ class Date extends React.Component {
    */
   renderDatePicker() {
     return (
-      this.state.open && <Portal >
+      this.state.open && <Portal parent={ this } onResposition={ this.updateDatePickerPosition.bind(this) }>
```
Comment (andrewjtait) [inline src/components/date/date.js]: does this need `.bind`?

File: src/components/date/date.js [hunk 3]
```diff
@@ -577,7 +567,7 @@ class Date extends React.Component {
    */
   renderDatePicker() {
     return (
-      this.state.open && <Portal >
+      this.state.open && <Portal onReposition={ this.updateDatePickerPosition.bind(this) }>
```
Comment (andrewjtait) [inline src/components/date/date.js]: does this need the `bind`?

File: src/components/date/date.js [hunk 4]
```diff
@@ -166,7 +166,6 @@ class Date extends React.Component {
     super(args);
     this.window = Browser.getWindow();
   }
-
```
Comment (andrewjtait) [inline src/components/date/date.js]: accidental removal of line?

File: src/components/dialog/__definition__.js [hunk 5]
```diff
@@ -21,7 +21,13 @@ const definition = new Definition('dialog', Dialog, {
...
   propValues: {
     title: 'Example Title for a Dialog',
-    children: '<Form>This is an example of a dialog.</Form>'
+    children: `<Form><DateInput name="date" label="Date" /><DateInput name="date" label="Date" /><DateInput name="date" label="Date" /><DateInput name="date" label="Date" />
+    <Textbox label="Name" validations={[ new PresenceValidation() ]}/>
+    <Textbox label="Name" validations={[ new PresenceValidation() ]}/>
+    <Textbox label="Name" validations={[ new PresenceValidation() ]}/>
+    <Textbox label="Name" validations={[ new PresenceValidation() ]}/>
+    <Textbox label="Name" validations={[ new PresenceValidation() ]}/>
+    <Textbox label="Name" validations={[ new PresenceValidation() ]}/>This is an example of a dialog.</Form>`
```
Comment (andrewjtait) [inline src/components/dialog/__definition__.js]: I'd either reset this, or tidy it up for the demo (give the inputs different names, make the it look a little organised)

File: src/components/portal/portal.js [hunk 6]
```diff
@@ -2,7 +2,7 @@ import React from 'react';
 import PropTypes from 'prop-types';
 import { createPortal } from 'react-dom';
 import Browser from '../../utils/helpers/browser';
-
+import ReactDOM from 'react-dom';
```
Comment (andrewjtait) [inline src/components/portal/portal.js]: this probably needs moving up in order, I'd guess eslint will complain about it. Also best keep the line break between the imports and the class.

File: src/components/portal/portal.js [hunk 7]
```diff
@@ -12,12 +12,34 @@ class Portal extends React.Component {
      * @property children
      * @type {Node}
      */
-    children: PropTypes.node
+    children: PropTypes.node,
+    onReposition: PropTypes.func
```
Comment (andrewjtait) [inline src/components/portal/portal.js]: Could add a description to this for what it is used for.

File: src/components/portal/portal.js [hunk 8]
```diff
@@ -12,12 +12,34 @@ class Portal extends React.Component {
...
   }
 
   componentWillUnmount() {
     Browser.getDocument().body.removeChild(this.defaultNode);
     this.defaultNode = null;
+    this.scrollParent && this.scrollParent.removeEventListener('scroll', this.props.onReposition);
+    Browser.getWindow().addEventListener('resize', this.props.onReposition);
```
Comment (andrewjtait) [inline src/components/portal/portal.js]: `addEventListener`? Think this should be `remove` ;)

Could wrap these in a `if (this.props.onReposition)` check too?

File: src/components/portal/portal.js [hunk 9]
```diff
@@ -12,12 +12,34 @@ class Portal extends React.Component {
...
+    if (this.props.onReposition) {
+      this.props.onReposition();
+      this.scrollParent = this.getScrollParent(ReactDOM.findDOMNode(this));
+      this.scrollParent && this.scrollParent.addEventListener('scroll', this.props.onReposition);
+      Browser.getWindow().addEventListener('resize', this.props.onReposition);
+    }
+  }
+
+  getScrollParent(element) {
+    const style = window.getComputedStyle(element);
```
Comment (andrewjtait) [inline src/components/portal/portal.js]: `Browser.getWindow()`

File: src/components/portal/portal.js [hunk 10]
```diff
@@ -12,12 +12,34 @@ class Portal extends React.Component {
...
+    if (style.position !== 'absolute' && /(auto|scroll)/.test(style.overflow + style.overflowY + style.overflowX)) {
+      return element;
+    }
+
+    return (element !== document.body) ? this.getScrollParent(element.parentElement) : null;
```
Comment (andrewjtait) [inline src/components/portal/portal.js]: `Browser.getDocument()`

File: src/components/dialog/__definition__.js (hunk 5 above)
Comment (JohnStuartBarrie) [reply src/components/dialog/__definition__.js]: Yes this was just done to give me a quick example for dev purposes. We will talk about the demo at the end as you suggested.

File: src/components/portal/__spec__.js [hunk 11]
```diff
@@ -67,7 +68,88 @@ describe('Portal', () => {
...
+      afterEach(() => {
+        wrapper.unmount();
+      });
+
+      it('will NOT add window "resize" listener ', () => {
+        expect(Browser.getDocument().addEventListener).not.toHaveBeenCalled();
+      });
+
+      it('will NOT add window "scroll" listener ', () => {
+        expect(Browser.getDocument().addEventListener).not.toHaveBeenCalled();
```
Comment (andrewjtait) [inline src/components/portal/__spec__.js]: is this not the same test as line 94? Maybe once is enough?

File: CHANGELOG.md [hunk 12]
```diff
@@ -83,6 +83,11 @@ npm install react@^16.2.0 react-dom@^16.2.0 --save
 ## React Portal
 
 We have updated the `Portal` component to use React's own version of portal which is available with React 16, removing the `react-portal` dependency.
+`Portal` now has an addition prop `onReposition` which is an optional callback function, called when the window resizes or a parent DOM element is scrolled.
```
Comment (andrewjtait) [inline CHANGELOG.md]: typo `additional`

File: src/components/portal/portal.js [hunk 13]
```diff
@@ -12,12 +12,43 @@ class Portal extends React.Component {
...
+    }
     Browser.getDocument().body.removeChild(this.defaultNode);
     this.defaultNode = null;
+    this.scrollParent = null;
+  }
+
+  getScrollParent(element) {
+    if (!element) { return null; }
+    const style = Browser.getWindow().getComputedStyle(element);
+    if (style && style.position !== 'absolute' &&
```
Comment (SeanArmstrong) [inline src/components/portal/portal.js]: Maybe we could give a name to one or two parts of this if statement to make it more readable

**Comment by iliaoni** [comment]
Fixed in Firefox, Safari, Chrome. QA passed.

**Pull request merged**